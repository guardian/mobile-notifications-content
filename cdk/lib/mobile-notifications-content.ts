import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { type App, CfnParameter, Duration, Fn } from 'aws-cdk-lib';
import {
	Alarm,
	ComparisonOperator,
	MathExpression,
	TreatMissingData,
} from 'aws-cdk-lib/aws-cloudwatch';
import { SnsAction } from 'aws-cdk-lib/aws-cloudwatch-actions';
import { Repository } from 'aws-cdk-lib/aws-ecr';
import { PolicyStatement, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import {
	DockerImageCode,
	DockerImageFunction,
	StartingPosition,
} from 'aws-cdk-lib/aws-lambda';
import { Topic } from 'aws-cdk-lib/aws-sns';

interface MobileNotificationsContentProps extends GuStackProps {
	crossAccountSsmRole: string;
	crossAccountDynamoRole: string;
	kinesisStreamArn: string;
	snsAlarmTopicArn: string;
}



export class MobileNotificationsContent extends GuStack {
	percentageAlarm(id: string, lamdba: DockerImageFunction, scope: MobileNotificationsContent) {
		const percentageErrorsContent = new MathExpression({
			expression: '100*m1/m2',
			usingMetrics: {
				m1: lamdba.metricErrors(),
				m2: lamdba.metricInvocations(),
			},
			label: `Error % of ${lamdba.functionName}`,
			period: Duration.minutes(10),
		});

		return new Alarm(scope, id, {
			alarmDescription: `High error percentage from ${lamdba.functionName} lambda in ${this.stage}`,
			comparisonOperator: ComparisonOperator.GREATER_THAN_THRESHOLD,
			evaluationPeriods: 1,
			threshold: 0,
			metric: percentageErrorsContent,
			treatMissingData: TreatMissingData.NOT_BREACHING,
		});
	}
	constructor(scope: App, id: string, props: MobileNotificationsContentProps) {
		super(scope, id, props);

		const appName = 'mobile-notifications-content';

		const contentImageRepositoryArn = Fn.importValue(
			'mobile-notifications-shared-resources-ContentLambdaContainerRepositoryArn',
		);

		const contentImageRepositoryName = Fn.importValue(
			'mobile-notifications-shared-resources-ContentLambdaContainerRepositoryUri',
		);

		const liveBlogsImageRepositoryArn = Fn.importValue(
			'mobile-notifications-shared-resources-LiveblogsLambdaContainerRepositoryArn',
		);

		const liveBlogsImageRepositoryName = Fn.importValue(
			'mobile-notifications-shared-resources-LiveblogsLambdaContainerRepositoryUri',
		);
		const buildId = new CfnParameter(this, 'BuildId', {
			type: 'String',
			default: 'dev',
			description: 'Tag to be used for the image URL, e.g. riff raff build id',
		}).value.toString();

		const executionRole = new Role(this, 'ExecutionRole', {
			assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
			path: '/',
		});

		executionRole.addToPolicy(
			new PolicyStatement({
				actions: ['sts:AssumeRole'],
				resources: [props.crossAccountSsmRole],
			}),
		);

		executionRole.addToPolicy(
			new PolicyStatement({
				actions: ['sts:AssumeRole'],
				resources: [props.crossAccountDynamoRole],
			}),
		);

		executionRole.addToPolicy(
			new PolicyStatement({
				actions: [
					'logs:CreateLogGroup',
					'logs:CreateLogStream',
					'logs:PutLogEvents',
				],
				resources: ['*'],
			}),
		);

		executionRole.addToPolicy(
			new PolicyStatement({
				actions: ['lambda:InvokeFunction'],
				resources: ['*'],
			}),
		);

		executionRole.addToPolicy(
			new PolicyStatement({
				actions: [
					'iam:PassRole',
					'iam:GenerateCredentialReport',
					'iam:Get*',
					'iam:List*',
				],
				resources: ['*'],
			}),
		);

		executionRole.addToPolicy(
			new PolicyStatement({
				actions: [
					'Kinesis:DescribeStream',
					'Kinesis:GetRecords',
					'Kinesis:GetShardIterator',
					'Kinesis:ListStream',
				],
				resources: ['*'],
			}),
		);

		executionRole.addToPolicy(
			new PolicyStatement({
				actions: ['cloudwatch:PutMetricData'],
				resources: ['*'],
			}),
		);

		const snsTopicAction = new SnsAction(
			Topic.fromTopicArn(this, 'AlarmTopic', props.snsAlarmTopicArn),
		);
		const contentLambda = new DockerImageFunction(this, 'ContentLambda', {
			functionName: `${appName}-${this.stage}-v2`,
			code: DockerImageCode.fromEcr(
				Repository.fromRepositoryAttributes(this, `${appName}-ecr`, {
					repositoryArn: contentImageRepositoryArn,
					repositoryName: contentImageRepositoryName,
				}),
				{
					tagOrDigest: buildId,
				},
			),
			environment: {
				Stage: this.stage,
				Stack: this.stack,
				App: appName,
				CrossAccountSsmReadingRole: `${props.crossAccountSsmRole}`,
			},
			memorySize: 4096,
			description: `Lambda to send notification when new content is published`,
			role: executionRole,
			timeout: Duration.seconds(60),
		});

		contentLambda.addEventSourceMapping('ContentLambdaEventMapping', {
			eventSourceArn: `${props.kinesisStreamArn}`,
			startingPosition: StartingPosition.LATEST,
			enabled: true,
			bisectBatchOnError: true,
		});
		const contentLambdaAlarm = this.percentageAlarm(`${appName}-alarm`, contentLambda, this)

		contentLambdaAlarm.addAlarmAction(snsTopicAction);
		contentLambdaAlarm.addOkAction(snsTopicAction);

		const liveBlogsLambda = new DockerImageFunction(this, 'LiveblogsLambda', {
			functionName: `${appName}-liveblogs-${this.stage}-v2`,
			code: DockerImageCode.fromEcr(
				Repository.fromRepositoryAttributes(this, `${appName}-liveblogs-ecr`, {
					repositoryArn: liveBlogsImageRepositoryArn,
					repositoryName: liveBlogsImageRepositoryName,
				}),
				{
					tagOrDigest: buildId,
				},
			),
			environment: {
				Stage: this.stage,
				Stack: this.stack,
				App: appName,
				CrossAccountSsmReadingRole: `${props.crossAccountSsmRole}`,
			},
			memorySize: 4096,
			description: `Lambda that sends push notifications when new key events are published on a liveblo`,
			role: executionRole,
			timeout: Duration.seconds(60),
		});

		liveBlogsLambda.addEventSourceMapping('LiveBlogsLambdaEventMapping', {
			eventSourceArn: `${props.kinesisStreamArn}`,
			startingPosition: StartingPosition.LATEST,
			enabled: true,
			bisectBatchOnError: true,
		});

		const liveBlogsLambdaAlarm = this.percentageAlarm(`${appName}-liveblogs-alarm`,liveBlogsLambda, this)
		liveBlogsLambdaAlarm.addAlarmAction(snsTopicAction);
		liveBlogsLambdaAlarm.addOkAction(snsTopicAction);
	}
}
