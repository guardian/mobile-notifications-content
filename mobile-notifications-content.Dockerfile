
FROM public.ecr.aws/lambda/java:11
COPY mobile-notifications-content.jar ${LAMBDA_TASK_ROOT}/lib/
CMD ["com.gu.mobile.content.notifications.ContentLambda::handler"]
