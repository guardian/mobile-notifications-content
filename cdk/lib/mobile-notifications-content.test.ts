import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { MobileNotificationsContent } from './mobile-notifications-content';

describe('The MobileNotificationsContent stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new MobileNotificationsContent(
			app,
			'MobileNotificationsContent',
			{
				stack: 'content-api-mobile-notifications',
				stage: 'TEST',
				crossAccountSsmRole:
					'arn:aws:iam::201359054765:role/mobile-content-notifications-lambda-cross-account-ssm-TEST',
				crossAccountDynamoRole:
					'arn:aws:iam::201359054765:role/mobile-content-notifications-lambda-cross-account-dynamo-TEST',
				kinesisStreamArn:
					'arn:aws:kinesis:eu-west-1:308506855511:stream/content-api-firehose-v2-TEST',
				snsAlarmTopicArn:
					'arn:aws:sns:eu-west-1:201359054765:mobile-server-side',
			},
		);
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
