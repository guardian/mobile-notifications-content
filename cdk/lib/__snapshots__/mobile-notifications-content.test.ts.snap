// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The MobileNotificationsContent stack matches the snapshot 1`] = `
{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Description": "lambda to replace the content-notifications-service",
  "Metadata": {
    "gu:cdk:constructs": [],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "App": {
      "Default": "mobile-notifications-content",
      "Description": "Application name",
      "Type": "String",
    },
    "BuildId": {
      "Default": "dev",
      "Description": "Tag to be used for the image URL, e.g. riff raff build id",
      "Type": "String",
    },
    "ConfigurationBucket": {
      "Default": "mobile-notifications-dist",
      "Description": "S3 Bucket containing app configuration",
      "Type": "String",
    },
    "CrossAccountBaseRoleName": {
      "Default": "mobile-content-notifications-lambda-cross-account",
      "Description": "Base constituent name of the mobile account roles to assume",
      "Type": "String",
    },
    "DeployBucket": {
      "Default": "content-api-dist",
      "Description": "S3 Bucket where RiffRaff uploads artifacts on deploy",
      "Type": "String",
    },
    "KinesisStreamName": {
      "Default": "content-api-firehose-v2",
      "Description": "Capi kinesis stream",
      "Type": "String",
    },
    "MobileAccountId": {
      "Default": 201359054765,
      "Description": "Account ID of the gu moble account which",
      "Type": "String",
    },
    "SharedResourcesStack": {
      "Default": "mobile-notifications-shared-resources",
      "Description": "Mobile notification content shared resources, e.g ecr repositories",
      "Type": "String",
    },
    "Stack": {
      "Default": "content-api-mobile-notifications",
      "Description": "Stack name",
      "Type": "String",
    },
    "Stage": {
      "AllowedValues": [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": {
    "ContentLambda9E34CFB0": {
      "DependsOn": [
        "ExecutionRoleDefaultPolicyA5B92313",
        "ExecutionRole605A040B",
      ],
      "Properties": {
        "Code": {
          "ImageUri": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    4,
                    {
                      "Fn::Split": [
                        ":",
                        {
                          "Fn::ImportValue": "mobile-notifications-shared-resources-ContentLambdaContainerRepositoryArn",
                        },
                      ],
                    },
                  ],
                },
                ".dkr.ecr.",
                {
                  "Fn::Select": [
                    3,
                    {
                      "Fn::Split": [
                        ":",
                        {
                          "Fn::ImportValue": "mobile-notifications-shared-resources-ContentLambdaContainerRepositoryArn",
                        },
                      ],
                    },
                  ],
                },
                ".",
                {
                  "Ref": "AWS::URLSuffix",
                },
                "/",
                {
                  "Fn::ImportValue": "mobile-notifications-shared-resources-ContentLambdaContainerRepositoryUri",
                },
                ":",
                {
                  "Ref": "BuildId",
                },
              ],
            ],
          },
        },
        "Description": "Lambda to send notification when new content is published",
        "Environment": {
          "Variables": {
            "App": "mobile-notifications-content",
            "CrossAccountSsmReadingRole": "arn:aws:iam::201359054765:role/mobile-content-notifications-lambda-cross-account-ssm-TEST",
            "Stack": "content-api-mobile-notifications",
            "Stage": "TEST",
          },
        },
        "FunctionName": "mobile-notifications-content-TEST-v2",
        "MemorySize": 4096,
        "PackageType": "Image",
        "Role": {
          "Fn::GetAtt": [
            "ExecutionRole605A040B",
            "Arn",
          ],
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-notifications-content",
          },
          {
            "Key": "Stack",
            "Value": "content-api-mobile-notifications",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ContentLambdaContentLambdaEventMappingA9509320": {
      "Properties": {
        "BisectBatchOnFunctionError": true,
        "Enabled": true,
        "EventSourceArn": "arn:aws:kinesis:eu-west-1:308506855511:stream/content-api-firehose-v2-TEST",
        "FunctionName": {
          "Ref": "ContentLambda9E34CFB0",
        },
        "StartingPosition": "LATEST",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "ExecutionRole": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": [
                  "lambda.amazonaws.com",
                ],
              },
            },
          ],
        },
        "Path": "/",
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": {
                "Action": "sts:AssumeRole",
                "Effect": "Allow",
                "Resource": {
                  "Fn::Sub": "arn:aws:iam::\${MobileAccountId}:role/\${CrossAccountBaseRoleName}-ssm-\${Stage}",
                },
              },
            },
            "PolicyName": "assume-mobile-ssm-read-config-role",
          },
          {
            "PolicyDocument": {
              "Statement": {
                "Action": "sts:AssumeRole",
                "Effect": "Allow",
                "Resource": {
                  "Fn::Sub": "arn:aws:iam::\${MobileAccountId}:role/\${CrossAccountBaseRoleName}-dynamo-\${Stage}",
                },
              },
            },
            "PolicyName": "assume-mobile-dynamo-role",
          },
          {
            "PolicyDocument": {
              "Statement": {
                "Action": [
                  "logs:CreateLogGroup",
                  "logs:CreateLogStream",
                  "logs:PutLogEvents",
                ],
                "Effect": "Allow",
                "Resource": "*",
              },
            },
            "PolicyName": "logs",
          },
          {
            "PolicyDocument": {
              "Statement": {
                "Action": [
                  "lambda:InvokeFunction",
                ],
                "Effect": "Allow",
                "Resource": "*",
              },
            },
            "PolicyName": "invoke-lambda",
          },
          {
            "PolicyDocument": {
              "Statement": {
                "Action": [
                  "iam:PassRole",
                  "iam:GenerateCredentialReport",
                  "iam:Get*",
                  "iam:List*",
                ],
                "Effect": "Allow",
                "Resource": "*",
              },
            },
            "PolicyName": "iam-read-and-assume",
          },
          {
            "PolicyDocument": {
              "Statement": {
                "Action": [
                  "Kinesis:DescribeStream",
                  "Kinesis:GetRecords",
                  "Kinesis:GetShardIterator",
                  "Kinesis:ListStream",
                ],
                "Effect": "Allow",
                "Resource": "*",
              },
            },
            "PolicyName": "kinesis-read",
          },
          {
            "PolicyDocument": {
              "Statement": {
                "Action": [
                  "cloudwatch:PutMetricData",
                ],
                "Effect": "Allow",
                "Resource": "*",
              },
            },
            "PolicyName": "cloudwatch-put-metric-daata",
          },
        ],
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-notifications-content",
          },
          {
            "Key": "Stack",
            "Value": "content-api-mobile-notifications",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ExecutionRole605A040B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Path": "/",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-notifications-content",
          },
          {
            "Key": "Stack",
            "Value": "content-api-mobile-notifications",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ExecutionRoleDefaultPolicyA5B92313": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": "arn:aws:iam::201359054765:role/mobile-content-notifications-lambda-cross-account-ssm-TEST",
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": "arn:aws:iam::201359054765:role/mobile-content-notifications-lambda-cross-account-dynamo-TEST",
            },
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "iam:PassRole",
                "iam:GenerateCredentialReport",
                "iam:Get*",
                "iam:List*",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "Kinesis:DescribeStream",
                "Kinesis:GetRecords",
                "Kinesis:GetShardIterator",
                "Kinesis:ListStream",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ExecutionRoleDefaultPolicyA5B92313",
        "Roles": [
          {
            "Ref": "ExecutionRole605A040B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LiveblogsLambda60F8E597": {
      "DependsOn": [
        "ExecutionRoleDefaultPolicyA5B92313",
        "ExecutionRole605A040B",
      ],
      "Properties": {
        "Code": {
          "ImageUri": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    4,
                    {
                      "Fn::Split": [
                        ":",
                        {
                          "Fn::ImportValue": "mobile-notifications-shared-resources-LiveblogsLambdaContainerRepositoryArn",
                        },
                      ],
                    },
                  ],
                },
                ".dkr.ecr.",
                {
                  "Fn::Select": [
                    3,
                    {
                      "Fn::Split": [
                        ":",
                        {
                          "Fn::ImportValue": "mobile-notifications-shared-resources-LiveblogsLambdaContainerRepositoryArn",
                        },
                      ],
                    },
                  ],
                },
                ".",
                {
                  "Ref": "AWS::URLSuffix",
                },
                "/",
                {
                  "Fn::ImportValue": "mobile-notifications-shared-resources-LiveblogsLambdaContainerRepositoryUri",
                },
                ":",
                {
                  "Ref": "BuildId",
                },
              ],
            ],
          },
        },
        "Description": "Lambda that sends push notifications when new key events are published on a liveblo",
        "Environment": {
          "Variables": {
            "App": "mobile-notifications-content",
            "CrossAccountSsmReadingRole": "arn:aws:iam::201359054765:role/mobile-content-notifications-lambda-cross-account-ssm-TEST",
            "Stack": "content-api-mobile-notifications",
            "Stage": "TEST",
          },
        },
        "FunctionName": "mobile-notifications-content-liveblogs-TEST-v2",
        "MemorySize": 4096,
        "PackageType": "Image",
        "Role": {
          "Fn::GetAtt": [
            "ExecutionRole605A040B",
            "Arn",
          ],
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-notifications-content",
          },
          {
            "Key": "Stack",
            "Value": "content-api-mobile-notifications",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LiveblogsLambdaLiveBlogsLambdaEventMapping6AF5A96F": {
      "Properties": {
        "BisectBatchOnFunctionError": true,
        "Enabled": true,
        "EventSourceArn": "arn:aws:kinesis:eu-west-1:308506855511:stream/content-api-firehose-v2-TEST",
        "FunctionName": {
          "Ref": "LiveblogsLambda60F8E597",
        },
        "StartingPosition": "LATEST",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "SenderErrorAlarm998AF4AD": {
      "Properties": {
        "AlarmActions": [
          "arn:aws:sns:eu-west-1:201359054765:mobile-server-side",
        ],
        "AlarmDescription": {
          "Fn::Join": [
            "",
            [
              "High error percentage from ",
              {
                "Ref": "LiveblogsLambda60F8E597",
              },
              " lambda in TEST",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "100*m1/m2",
            "Id": "expr_1",
            "Label": {
              "Fn::Join": [
                "",
                [
                  "Error % of ",
                  {
                    "Ref": "LiveblogsLambda60F8E597",
                  },
                ],
              ],
            },
          },
          {
            "Id": "m1",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "FunctionName",
                    "Value": {
                      "Ref": "LiveblogsLambda60F8E597",
                    },
                  },
                ],
                "MetricName": "Errors",
                "Namespace": "AWS/Lambda",
              },
              "Period": 600,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "m2",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "FunctionName",
                    "Value": {
                      "Ref": "LiveblogsLambda60F8E597",
                    },
                  },
                ],
                "MetricName": "Invocations",
                "Namespace": "AWS/Lambda",
              },
              "Period": 600,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "OKActions": [
          "arn:aws:sns:eu-west-1:201359054765:mobile-server-side",
        ],
        "Threshold": 0,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
}
`;
